src/Shared/WeakStringCache.cs

```diff
@@ -0,0 +1,241 @@
+            /// </summary>
+            /// <param name="internable">The internable describing the string we're looking for.</param>
+            /// <returns>The string matching the internable or null if the handle is referencing a collected string or the string is different.</returns>
+            public string GetString<T>(T internable) where T : IInternable
+            {
+                if (WeakHandle.IsAllocated && WeakHandle.Target is string str)
+                {
+                    if (internable.Length == str.Length &&
+                        internable.StartsWithStringByOrdinalComparison(str))
+                    {
+                        return str;
+                    }
+                }
+                return null;
+            }
+
+            /// <summary>
+            /// Sets the handle to the given string. If the handle is still referencing another live string, it is effectively forgotten.
+            /// </summary>
+            /// <param name="str">The string to add.</param>
+            public void SetString(string str)
+            {
+                if (!WeakHandle.IsAllocated)
+                {
+                    // The handle is not allocated - allocate it.
+                    WeakHandle = GCHandle.Alloc(str, GCHandleType.Weak);
+                }
+                else
+                {
+                    WeakHandle.Target = str;
+                }
+            }
+
+            /// <summary>
+            /// Frees the GC handle.
+            /// </summary>
+            public void Free()
+            {
+                WeakHandle.Free();
+            }
+        }
+
+        /// <summary>
+        /// Improvised capacity for the scavenging heuristics.
+        /// </summary>
+        private int _capacity = 100;
+
+        /// <summary>
+        /// A dictionary mapping string hash codes to weak GC handles.
+        /// </summary>
+        private Dictionary<int, StringWeakHandle> _stringsByHashCode;
+
+        public WeakStringCache()
+        {
+            _stringsByHashCode = new Dictionary<int, StringWeakHandle>(_capacity);
+        }
+
+        /// <summary>
+        /// Main entrypoint of this cache. Tries to look up a string that matches the given internable. If it succeeds, returns
+        /// the string and sets cacheHit to true. If the string is not found, calls ExpensiveConvertToString on the internable,
+        /// adds the resulting string to the cache, and returns it setting cacheHit to false.
+        /// </summary>
+        /// <param name="internable">The internable describing the string we're looking for.</param>
+        /// <returns>A string matching the given internable.</returns>
+        public string GetOrCreateEntry<T>(T internable, out bool cacheHit) where T : IInternable
+        {
+            int hashCode = GetInternableHashCode(internable);
+
+            StringWeakHandle handle;
+            string result;
+            lock (_stringsByHashCode)
```

Comment:
Quickbuild does parallel evaluations. As far as I remember, lock contention in the opportunistic string cache was showing as a bottleneck so they added the ConcurrentDictionary implementation to relieve it. This means QB will most likely have to keep using the ConcurrentDictionary approach.

Reply:
Actually, you could use a ConcurrentDictionary and only fully lock when scavenging right?

---
In reply to: [428400003](https://github.com/microsoft/msbuild/pull/5353#discussion_r428400003) [](ancestors = 428400003)

Reply:
I was considering `ConcurrentDictionary` but it's not available on downlevel so it didn't seem worth it. If parallel evaluations are a thing, it definitely changes the priorities, thank you!

I have switched `WeakStringCache` to use `ConcurrentDictionary` where available. Not relying on `AddOrUpdate` because it doesn't have much benefits in this use case and has about the same perf as `TryGetValue` followed by `TryAdd`. I found places where we could leak handles, though, and fixed them.

How difficult would it be to test these changes in Quickbuild before or shortly after merging?

Reply:
`ConcurrentDictionary` is available back to .NET Framework 4.5. How far do you guys support downlevel?

https://apisof.net/catalog/System.Collections.Concurrent.ConcurrentDictionary%3CTKey,TValue%3E

Reply:
.NET 3.5 for MSBuildTaskHost :(

I'm happy to have "correct enough but horrible performance" implementations for that codepath, though. That's what I'm doing for `System.Collections.Immutable` in #5279.